### PR TITLE
Add support for Hardware Key PIN

### PR DESCRIPTION
**Pull request by Joerger**
I recommend reviewing commit by commit.

Changes:
- Update RFD 80 to cover changes below
- Simplify policy/mfa logic to make the changes below smaller
- Add require MFA types `hardware_key_pin` and `hardware_key_touch`. Like the existing `hardware_key_touch`, these types count as a replacement for MFA verification for per-session MFA and admin actions MFA.
- Add support for PIV pin policies:
  - Prompt for PIN when needed
  - Require user pins to be non-default - prompt for new PIN when the default is given by the user.

Based off of https://github.com/gravitational/teleport/pull/32275

### CODE REVIEW

**Comment by Joerger** [comment]
@rosstimothy @capnspacehook friendly ping to review

**Comment by codingllama** [comment]
How are doing we on the base branches? I see a few jumps between this and master, some of which seem to be drafts yet?

**Comment by Joerger** [comment]
> How are doing we on the base branches? I see a few jumps between this and master, some of which seem to be drafts yet?

Development is done, this is just based off of:
 - [piv-slot PR](https://github.com/gravitational/teleport/pull/31732)
 - [piv-slot e PR](https://github.com/gravitational/teleport.e/pull/2191)
 - [piv-slot e ref update / deprecation follow up](https://github.com/gravitational/teleport/pull/32275)

Ideally those 3 PRs would just be one PR, but alas.

File: rfd/0080-hardware-key-support.md
Comment (codingllama) [inline rfd/0080-hardware-key-support.md:447]: > handled by a standard IO prompt

Meaning a password style prompt, correct?

**Review by codingllama — COMMENTED** [review]
Partial review, more later.

File: rfd/0080-hardware-key-support.md
Comment (Joerger) [reply rfd/0080-hardware-key-support.md:447]: Yes

**Review by jentfoo — APPROVED** [review]

**Review by jentfoo — APPROVED** [review]

File: api/types/authentication.go
Comment (codingllama) [inline api/types/authentication.go:870]: nit: Missing godoc.

File: api/types/authentication.go
Comment (codingllama) [inline api/types/authentication.go:925]: nit: godocs

File: api/types/authentication.go
Comment (codingllama) [inline api/types/authentication.go:967]: Is the decode/encode logic worth testing?

File: api/utils/keys/policy.go
Comment (codingllama) [inline api/utils/keys/policy.go:121]: Maybe: PolicyThatSatisfiesSet ?

File: api/utils/keys/policy.go
Comment (codingllama) [inline api/utils/keys/policy.go:129]: Is this ever false for the set of policies we have today?

File: api/utils/keys/policy.go
Comment (codingllama) [inline api/utils/keys/policy.go:135]: Suggestion: remove indent, replace if/else chain with switch, line breaks here and there.

```suggestion
		if IsRequiredPolicyMet(policy, setPolicy) {
			continue
		}

		switch {
		case IsRequiredPolicyMet(setPolicy, policy):
			// Upgrade set policy to stricter policy.
			setPolicy = policy
		case IsRequiredPolicyMet(policy, PrivateKeyPolicyHardwareKeyTouchAndPIN) &&
				IsRequiredPolicyMet(setPolicy, PrivateKeyPolicyHardwareKeyTouchAndPIN):
			// Neither policy is met by the other (pin or touch), but both are met by
			// stricter pin+touch policy.
			setPolicy = PrivateKeyPolicyHardwareKeyTouchAndPIN
		default:
			return PrivateKeyPolicyNone, trace.BadParameter("" +
					"private key policy requirements %q and %q are incompatible, " + 
					"please contact the cluster administrator", policy, setPolicy)
		}
```

File: api/utils/keys/policy_test.go
Comment (codingllama) [inline api/utils/keys/policy_test.go:66]:
```suggestion
		matchingPolicies []keys.PrivateKeyPolicy
```

Assuming I read it correctly?

File: api/utils/keys/policy_test.go
Comment (codingllama) [inline api/utils/keys/policy_test.go:101]: nit: wantPolicy, in keeping with Go got/want lingo?

```suggestion
		wantPolicy keys.PrivateKeyPolicy
```

File: api/utils/keys/policy_test.go
Comment (codingllama) [inline api/utils/keys/policy_test.go:157]:
```suggestion
			slices.Reverse(tc.policySet)
```

If you want to backport, there's also /x/exp/slices.Reverse.

File: api/utils/keys/policy_test.go
Comment (codingllama) [inline api/utils/keys/policy_test.go:166]: What's the method being tested here? Could we name the test after it?

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:258]: If the message is only checked here, I'd move the const here. I'd also add a comment that it was acquired from an actual PIV key.

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:53]: This makes me a bit uneasy, matching strings is usually brittle. Is there something better we could do?

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:377]: Suggestion:

```go
  isPINPolicy := ...
  isTouchPolicy := ...
  switch {
  case isPINPolicy && isTouchPolicy:
    // ...
  case isPINPolicy:
    // ...
  case isTouchPolicy:
    // ...
  default:
    // ...
  }
```

Simpler to read than nested switches.

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:550]: Does this need some sort of user interaction?

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:566]:
```suggestion
	pin, err := prompt.Password(ctx, os.Stderr, prompt.Stdin(), "Enter your YubiKey PIV PIN [blank to use default PIN]")
```

Is that correct? I don't totally follow the prompt.

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:778]: Why the numeric restriction? Do we really have to check this?

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:774]: Why the max length restriction?

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:792]:
```suggestion
		newPINConfirm, err := prompt.Password(ctx, os.Stderr, prompt.Stdin(), "Confirm your new YubiKey PIV PIN")
```

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:808]:
```suggestion
			fmt.Fprintf(os.Stderr, "PIN must be numeric and between 6 to 8 digits.\n")
```

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:827]: Same suggestions for prompts here.

**Review by codingllama — COMMENTED** [review]
Partial review, more to follow later.

File: api/utils/keys/policy.go
Comment (Joerger) [reply api/utils/keys/policy.go:129]: No, but this additional check was suggested in a previous comment to future proof us for future key policies. For example, if we ever introduce `web_session`  and `hardware_key_strict` (which would prevent web sessions without PIV) as possible requirements.

File: api/utils/keys/policy.go
Comment (Joerger) [reply api/utils/keys/policy.go:121]: Wow, you have no clue how long I've been looking for the word "satisfied" to replace "met". Too bad I can't change the key policy error from "private key policy not met" to "private key policy not satisfied" due to the client dependency...

File: api/utils/keys/yubikey.go
Comment (Joerger) [reply api/utils/keys/yubikey.go:53]: The actual error code returned by the smart card is "6982", and the PIV library wraps that with the user readable message. Unfortunately the PIV library doesn't export any of the methods used. Would it be better to just match on "6982"?

File: api/utils/keys/yubikey.go
Comment (Joerger) [reply api/utils/keys/yubikey.go:550]: No, it just needs the correct old PIN, and only has 3 attempts before the card is locked.

File: api/utils/keys/yubikey.go
Comment (Joerger) [reply api/utils/keys/yubikey.go:566]: Yes that works too. I'll update the RFD to show the prompt flow, any suggestions are welcome. It's a bit of a pain to have the user set 2 PINs but there isn't really an alternative that I can think of.
```bash
> tsh login --user=dev
Enter your YubiKey PIV PIN [blank to use default PIN]:
# \n
The default PIN 123456 is not supported.
Please set a new 6 digit PIN:
Enter your new YubiKey PIV PIN:
Enter your new YubiKey PIV PIN again to confirm:
Enter your YubiKey PIV PUK to reset PIN [blank to use default PUK]
# \n
The default PUK 12345678 is not supported
Please set a new 8 digit PUK:
Enter your new YubiKey PIV PUK:
Enter your new YubiKey PIV PUK again to confirm:
```

File: api/utils/keys/yubikey.go
Comment (Joerger) [reply api/utils/keys/yubikey.go:774]:
```
ykman piv access change-pin
Enter the current PIN: 
Enter the new PIN: 
Repeat for confirmation: 
Usage: ykman piv access change-pin [OPTIONS]
Try 'ykman piv access change-pin -h' for help.

Error: New PIN must be between 6 and 8 characters long.
```

File: api/utils/keys/policy.go
Comment (codingllama) [reply api/utils/keys/policy.go]: I'd add a small comment, but otherwise seems fine to me.

File: api/utils/keys/yubikey.go
Comment (codingllama) [reply api/utils/keys/yubikey.go]: I considered that. Maybe split on `[\s:]` and see if the word is there? It's just an idea though.

File: api/utils/keys/yubikey.go
Comment (codingllama) [reply api/utils/keys/yubikey.go:552]: Ta. I'd say in the docs that it doesn't need user interaction, just to be extra clear.

File: api/utils/keys/yubikey.go
Comment (codingllama) [reply api/utils/keys/yubikey.go]: Your call, I mostly wanted to clarify. FYI I don't usually keep updating RFDs for minor changes like user messages (nor do I go back after a certain time has passed).

File: api/utils/keys/yubikey.go
Comment (codingllama) [reply api/utils/keys/yubikey.go]: Huh, TIL. It's a Yubico limitation then? I'd mention that in a comment, seems arbitrarily small otherwise.

File: api/utils/keys/policy.go
Comment (codingllama) [inline api/utils/keys/policy.go:150]: You could solve this with a single regex:

```suggestion
	privateKeyPolicyErrRegex = regexp.MustCompile(`private key policy not (met|satisfied): (\w+)`)
```

Just make sure to update the number of subMatches/policy group index below.

File: api/utils/keys/policy_test.go
Comment (codingllama) [inline api/utils/keys/policy_test.go:165]:
```suggestion
// TestParsePrivateKeyPolicyError tests private key policy error parsing and checking.
```

File: api/utils/keys/yubikey.go
Comment (codingllama) [inline api/utils/keys/yubikey.go:255]: nit: 0x6982 is not the same as decimal 6982, is that correct?

File: api/utils/keys/yubikey.go
Comment (codingllama) [reply api/utils/keys/yubikey.go:552]: (By "docs" I meant godoc.)

File: e
Comment (codingllama) [inline e:1]: Revert this? e/ is already past your commit.

**Review by codingllama — APPROVED** [review]
Looks good. All done now, apologies for the delays!

I'm out tomorrow so I'll leave it pre-approved, but please address the remaining comments before merging.

File: api/utils/keys/yubikey.go
Comment (Joerger) [reply api/utils/keys/yubikey.go]: > split on [\s:]

What does this mean?

File: api/utils/keys/yubikey.go
Comment (Joerger) [reply api/utils/keys/yubikey.go]: Yeah, I wasn't sure what to do. Seems like some things are worth adding back to the RFD but probably not this. I'll just add it to the docs next time.

File: api/utils/keys/yubikey.go
Comment (Joerger) [reply api/utils/keys/yubikey.go:255]: Yes, The PIV library deals with it as a hexadecimal `0x6982` but ultimately stringifies it as "6982" in the error message.

File: api/utils/keys/yubikey.go
Comment (codingllama) [reply api/utils/keys/yubikey.go]: Sorry, that was a bit cryptic. I meant split on spaces or `:`, as in: 

```go
re := regexp.MustCompile(`[\s:]`)
if slices.Contains(re.Split(err.Error(), "6982") {
  // ...
}
```

Comment (Joerger) [reply api/utils/keys/yubikey.go]: Ah I see, I think just depending on the error code and no error structure will be the least brittle, we can require the error structure later if it's needed.

File: api/utils/keys/yubikey.go
Comment (codingllama) [reply api/utils/keys/yubikey.go:255]: I think that's worth calling out in the comment, as it seems like an error liable to be fixed by them (probably by printing 0x6982, I'd imagine, but it could be translated to decimal too).

Comment (Joerger) [reply api/utils/keys/yubikey.go:255]: They're doing `fmt.Sprintf("smart card error %04x%s", 0x6982, "security status not satisfied")`, so it looks intentional.

**Comment by Joerger** [comment]
@r0mant Can you +1 this one?

File: api/utils/keys/yubikey_test.go
Comment (codingllama) [inline api/utils/keys/yubikey_test.go:54]:
```suggestion
	prompt.SetStdin(prompt.NewFakeReader().AddString(testPIN))
```

and drop mockPinReader.

Comment (codingllama) [reply api/utils/keys/yubikey_test.go]: This is fine, it represents the test anyway. Thanks.

**Review by r0mant — APPROVED** [review]
Bot.

**Pull request merged**
